Add SAD test.

RyanZhao-Cidana opened the pull request.

edmond-zhu committed:
Add CFL subtract average test

1. Add test for CFL subtract average functions

suanli committed:
Add SAD test

Add unit test for sad functions

Cidana-Zhengwen commented on test/SadTest.cc:
```diff
@@ -0,0 +1,500 @@
+        uint32_t ref_sad = 0;
+        uint32_t non_avx2_sad = 0;
+        uint32_t avx2_sad = 0;
+        ref_sad = reference_sad();
```
Is it typo? reference_sad => reference_sad_avg ?

Cidana-Zhengwen commented on test/SadTest.cc:
```diff
@@ -0,0 +1,500 @@
+        source_stride_ = ref_1_stride_ = ref_2_stride_ = width_ * 2;
+
+        int32_t mask = (1 << 8) - 1;
+        SVTRandom *rnd = new SVTRandom(0, mask);
```
It is better move this big switch case out of constructor.

Cidana-Zhengwen commented on test/SadTest.cc:
```diff
@@ -0,0 +1,500 @@
+    void check_sad() {
+        EB_SADKERNELNxM_TYPE non_avx2_func = nullptr;
+        EB_SADKERNELNxM_TYPE avx2_func = nullptr;
+        switch (width_) {
```
Maybe we can optimize this big switch case, such as:
non_avx2_func =NxMSadKernelSubSampled_funcPtrArray[ASM_NON_AVX2][width_>>3];
avx2_func = NxMSadKernelSubSampled_funcPtrArray[ASM_AVX2][width_>>3];

Cidana-Zhengwen commented on test/SadTest.cc:
```diff
@@ -0,0 +1,500 @@
+    void check_sad() {
+        EB_SADKERNELNxM_TYPE non_avx2_func = nullptr;
+        EB_SADKERNELNxM_TYPE avx2_func = nullptr;
+        switch (width_) {
```
Maybe we can optimize this big switch case by funcPtrArray[XXX][width_>>3];

Cidana-Zhengwen commented on test/SadTest.cc:
```diff
@@ -0,0 +1,500 @@
+    void check_sad_avg() {
+        EB_SADAVGKERNELNxM_TYPE non_avx2_func = nullptr;
+        EB_SADAVGKERNELNxM_TYPE avx2_func = nullptr;
+        switch (width_) {
```
Maybe we can optimize this big switch case by funcPtrArray[XXX][width_>>3];

Cidana-Zhengwen reviewed (changes requested):
These 3 types of tests have common test codes, it is better combine these 3 types into single test class.
We may add enum TestType, and use common function pointer type for 3 types of non_avx_func and avx_func .

edmond-zhu committed:
Add unit tests of compute mean functions

1. The functions include mean and mean of squares.

Cidana-Developers committed:
Refine sad test

1. reuse SADTestBase for GetEight Test and Loop Test;
2. fix some parameter errors
